File: todo/api-php/endpoints/lists.php
<?php
/**
 * Endpoints de Listas
 * Todo App - API PHP
 */

function handleLists(?string $id, string $method): void {
    $auth = requireAuth();
    
    switch ($method) {
        case 'GET':
            if ($id) {
                getList($auth['user_id'], $id);
            } else {
                getLists($auth['user_id']);
            }
            break;
        case 'POST':
            createList($auth['user_id']);
            break;
        case 'PUT':
            if ($id) updateList($auth['user_id'], $id);
            break;
        case 'DELETE':
            if ($id) deleteList($auth['user_id'], $id);
            break;
        default:
            jsonResponse(false, 'Método não permitido', null, 405);
    }
}

function getLists(string $userId): void {
    $pdo = getDB();
    
    $stmt = $pdo->prepare("
        SELECT id, user_id, name, color, created_at, updated_at
        FROM todo_lists 
        WHERE user_id = ? AND deleted_at IS NULL
        ORDER BY created_at DESC
    ");
    $stmt->execute([$userId]);
    $lists = $stmt->fetchAll();
    
    jsonResponse(true, 'Listas carregadas', $lists);
}

function getList(string $userId, string $listId): void {
    $pdo = getDB();
    
    $stmt = $pdo->prepare("
        SELECT id, user_id, name, color, created_at, updated_at
        FROM todo_lists 
        WHERE id = ? AND user_id = ? AND deleted_at IS NULL
    ");
    $stmt->execute([$listId, $userId]);
    $list = $stmt->fetch();
    
    if (!$list) {
        jsonResponse(false, 'Lista não encontrada', null, 404);
    }
    
    jsonResponse(true, 'Lista encontrada', $list);
}

function createList(string $userId): void {
    $data = getJsonInput();
    
    $id = $data['id'] ?? generateUUID();
    $name = trim($data['name'] ?? '');
    $color = $data['color'] ?? '#3B82F6';
    $createdAt = $data['created_at'] ?? date('Y-m-d H:i:s');
    
    if (empty($name)) {
        jsonResponse(false, 'Nome da lista é obrigatório', null, 400);
    }
    
    $pdo = getDB();
    
    // Verificar se já existe (para sincronização)
    $stmt = $pdo->prepare("SELECT id FROM todo_lists WHERE id = ?");
    $stmt->execute([$id]);
    
    if ($stmt->fetch()) {
        // Atualizar se já existe
        $stmt = $pdo->prepare("
            UPDATE todo_lists 
            SET name = ?, color = ?, updated_at = NOW(), deleted_at = NULL
            WHERE id = ? AND user_id = ?
        ");
        $stmt->execute([$name, $color, $id, $userId]);
    } else {
        // Criar nova
        $stmt = $pdo->prepare("
            INSERT INTO todo_lists (id, user_id, name, color, created_at) 
            VALUES (?, ?, ?, ?, ?)
        ");
        $stmt->execute([$id, $userId, $name, $color, $createdAt]);
    }
    
    // Log de sincronização
    logSync($userId, 'list', $id, 'create');
    
    jsonResponse(true, 'Lista criada com sucesso', [
        'id' => $id,
        'user_id' => $userId,
        'name' => $name,
        'color' => $color,
        'created_at' => $createdAt
    ], 201);
}

function updateList(string $userId, string $listId): void {
    $data = getJsonInput();
    
    $name = trim($data['name'] ?? '');
    $color = $data['color'] ?? null;
    
    $pdo = getDB();
    
    // Verificar propriedade
    $stmt = $pdo->prepare("SELECT id FROM todo_lists WHERE id = ? AND user_id = ? AND deleted_at IS NULL");
    $stmt->execute([$listId, $userId]);
    
    if (!$stmt->fetch()) {
        jsonResponse(false, 'Lista não encontrada', null, 404);
    }
    
    $updates = [];
    $params = [];
    
    if (!empty($name)) {
        $updates[] = "name = ?";
        $params[] = $name;
    }
    
    if ($color !== null) {
        $updates[] = "color = ?";
        $params[] = $color;
    }
    
    if (empty($updates)) {
        jsonResponse(false, 'Nenhum campo para atualizar', null, 400);
    }
    
    $updates[] = "updated_at = NOW()";
    $params[] = $listId;
    $params[] = $userId;
    
    $sql = "UPDATE todo_lists SET " . implode(', ', $updates) . " WHERE id = ? AND user_id = ?";
    $stmt = $pdo->prepare($sql);
    $stmt->execute($params);
    
    // Log de sincronização
    logSync($userId, 'list', $listId, 'update');
    
    // Retornar lista atualizada
    getList($userId, $listId);
}

function deleteList(string $userId, string $listId): void {
    $pdo = getDB();
    
    // Soft delete
    $stmt = $pdo->prepare("
        UPDATE todo_lists 
        SET deleted_at = NOW() 
        WHERE id = ? AND user_id = ?
    ");
    $stmt->execute([$listId, $userId]);
    
    // Soft delete das tarefas
    $stmt = $pdo->prepare("
        UPDATE tasks 
        SET deleted_at = NOW() 
        WHERE list_id = ?
    ");
    $stmt->execute([$listId]);
    
    // Log de sincronização
    logSync($userId, 'list', $listId, 'delete');
    
    jsonResponse(true, 'Lista deletada com sucesso');
}

function generateUUID(): string {
    return sprintf(
        '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
        mt_rand(0, 0xffff), mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0x0fff) | 0x4000,
        mt_rand(0, 0x3fff) | 0x8000,
        mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
    );
}

function logSync(string $userId, string $entityType, string $entityId, string $action): void {
    $pdo = getDB();
    $stmt = $pdo->prepare("
        INSERT INTO sync_log (user_id, entity_type, entity_id, action) 
        VALUES (?, ?, ?, ?)
    ");
    $stmt->execute([$userId, $entityType, $entityId, $action]);
}

File: todo-api-rust/src/handlers/lists.rs
use actix_web::{web, HttpRequest, HttpResponse};
use chrono::Utc;
use sqlx::MySqlPool;

use crate::config::Config;
use crate::errors::{ApiError, ApiResponse};
use crate::middleware::jwt::get_auth_user;
use crate::models::*;

// Helper: Generate UUID
fn generate_uuid() -> String {
    uuid::Uuid::new_v4().to_string()
}

// Helper: Log sync
async fn log_sync(pool: &MySqlPool, user_id: &str, entity_type: &str, entity_id: &str, action: &str) {
    let _ = sqlx::query(
        "INSERT INTO sync_log (user_id, entity_type, entity_id, action) VALUES (?, ?, ?, ?)"
    )
    .bind(user_id)
    .bind(entity_type)
    .bind(entity_id)
    .bind(action)
    .execute(pool)
    .await;
}

// GET /lists
pub async fn get_lists(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;

    let lists: Vec<TodoList> = sqlx::query_as(
        r#"
        SELECT id, user_id, name, color, created_at, updated_at, deleted_at
        FROM todo_lists 
        WHERE user_id = ? AND deleted_at IS NULL
        ORDER BY created_at DESC
        "#
    )
    .bind(&claims.user_id)
    .fetch_all(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let lists: Vec<ListResponse> = lists.into_iter().map(|l| l.into()).collect();

    Ok(HttpResponse::Ok().json(ApiResponse::success("Listas carregadas", lists)))
}

// GET /lists/{id}
pub async fn get_list(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let list_id = path.into_inner();

    let list: Option<TodoList> = sqlx::query_as(
        r#"
        SELECT id, user_id, name, color, created_at, updated_at, deleted_at
        FROM todo_lists 
        WHERE id = ? AND user_id = ? AND deleted_at IS NULL
        "#
    )
    .bind(&list_id)
    .bind(&claims.user_id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let list = list.ok_or_else(|| ApiError::not_found("Lista não encontrada"))?;

    Ok(HttpResponse::Ok().json(ApiResponse::success("Lista encontrada", ListResponse::from(list))))
}

// POST /lists
pub async fn create_list(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    body: web::Json<CreateListRequest>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;

    let id = body.id.clone().unwrap_or_else(generate_uuid);
    let name = body.name.trim();
    let color = body.color.clone().unwrap_or_else(|| "#3B82F6".to_string());
    let created_at = body.created_at.clone().unwrap_or_else(|| Utc::now().to_rfc3339());

    if name.is_empty() {
        return Err(ApiError::bad_request("Nome da lista é obrigatório"));
    }

    // Check if exists (for sync)
    let existing: Option<(String,)> = sqlx::query_as(
        "SELECT id FROM todo_lists WHERE id = ?"
    )
    .bind(&id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    if existing.is_some() {
        // Update if exists
        sqlx::query(
            r#"
            UPDATE todo_lists 
            SET name = ?, color = ?, updated_at = NOW(), deleted_at = NULL
            WHERE id = ? AND user_id = ?
            "#
        )
        .bind(name)
        .bind(&color)
        .bind(&id)
        .bind(&claims.user_id)
        .execute(pool.get_ref())
        .await
        .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;
    } else {
        // Create new
        sqlx::query(
            r#"
            INSERT INTO todo_lists (id, user_id, name, color, created_at) 
            VALUES (?, ?, ?, ?, ?)
            "#
        )
        .bind(&id)
        .bind(&claims.user_id)
        .bind(name)
        .bind(&color)
        .bind(&created_at)
        .execute(pool.get_ref())
        .await
        .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;
    }

    log_sync(pool.get_ref(), &claims.user_id, "list", &id, "create").await;

    Ok(HttpResponse::Created().json(ApiResponse::success(
        "Lista criada com sucesso",
        ListResponse {
            id,
            user_id: claims.user_id,
            name: name.to_string(),
            color,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        },
    )))
}

// PUT /lists/{id}
pub async fn update_list(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
    body: web::Json<UpdateListRequest>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let list_id = path.into_inner();

    // Check ownership
    let existing: Option<TodoList> = sqlx::query_as(
        "SELECT * FROM todo_lists WHERE id = ? AND user_id = ? AND deleted_at IS NULL"
    )
    .bind(&list_id)
    .bind(&claims.user_id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let mut list = existing.ok_or_else(|| ApiError::not_found("Lista não encontrada"))?;

    // Update fields
    let mut has_updates = false;

    if let Some(name) = &body.name {
        let name = name.trim();
        if !name.is_empty() {
            list.name = name.to_string();
            has_updates = true;
        }
    }

    if let Some(color) = &body.color {
        list.color = color.clone();
        has_updates = true;
    }

    if !has_updates {
        return Err(ApiError::bad_request("Nenhum campo para atualizar"));
    }

    sqlx::query(
        "UPDATE todo_lists SET name = ?, color = ?, updated_at = NOW() WHERE id = ? AND user_id = ?"
    )
    .bind(&list.name)
    .bind(&list.color)
    .bind(&list_id)
    .bind(&claims.user_id)
    .execute(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    log_sync(pool.get_ref(), &claims.user_id, "list", &list_id, "update").await;

    // Return updated list
    let list: TodoList = sqlx::query_as(
        "SELECT * FROM todo_lists WHERE id = ?"
    )
    .bind(&list_id)
    .fetch_one(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    Ok(HttpResponse::Ok().json(ApiResponse::success("Lista atualizada", ListResponse::from(list))))
}

// DELETE /lists/{id}
pub async fn delete_list(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let list_id = path.into_inner();

    // Soft delete list
    sqlx::query(
        "UPDATE todo_lists SET deleted_at = NOW() WHERE id = ? AND user_id = ?"
    )
    .bind(&list_id)
    .bind(&claims.user_id)
    .execute(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    // Soft delete tasks
    sqlx::query(
        "UPDATE tasks SET deleted_at = NOW() WHERE list_id = ?"
    )
    .bind(&list_id)
    .execute(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    log_sync(pool.get_ref(), &claims.user_id, "list", &list_id, "delete").await;

    Ok(HttpResponse::Ok().json(ApiResponse::<()>::success_no_data("Lista deletada com sucesso")))
}
